### ISSUE TITLE
Should the Engine iterate over and consume IEnumerable and IQueryable results?

### ISSUE THREAD

**Issue by adamsitnik** [dotnet/BenchmarkDotNet#858]
Today I have discovered that one of the ML.NET benchmarks returns `IEnumerable` and takes 16 nanoseconds to execute. It was suspicious to me, so I decided to iterate over the result and consume the result.

```cs
[Benchmark]
public IEnumerable<IrisPrediction> Before() => s_trainedModel.Predict(s_batches[0]);

private readonly Consumer _consumer = new Consumer();

[Benchmark]
public void After() => Consume(s_trainedModel.Predict(s_batches[0]));

private void Consume(IEnumerable<IrisPrediction> predictions)
{
    foreach (var prediction in predictions)
        _consumer.Consume(prediction);
}
```

It turned out that after we consume it, it takes 2.370 ms, not 16 ns.

I would like to start a dicussion here: should BenchmarkDotNet handle IQueryable results and execute them?

@AndreyAkinshin @jorive what do you think?

**Comment by redknightlois** [comment]
Makes sense. It is also a common error when benchmarking. I am not saying I have been guilty of it, BUT if you try to guess probably you will be right :D.

**Comment by adamsitnik** [comment]
I had an offline conversation with @jorive and we agreed that most probably a validator with error + a public extension method for consuming the result would be best

**Comment by adamsitnik** [comment]
I added a new Validator which gives a clear error message with explanation + solution. With samples and more info in the docs

![image](https://user-images.githubusercontent.com/6011991/44474732-a06f6d00-a633-11e8-883c-7188430e79ba.png)

**Comment by AndreyAkinshin** [comment]
@adamsitnik, looks awesome!
